duplex_http_call.go

```diff
@@ -77,17 +78,31 @@ func newDuplexHTTPCall(
 		Proto:      "HTTP/1.1",
 		ProtoMajor: 1,
 		ProtoMinor: 1,
-		Body:       pipeReader,
 		Host:       url.Host,
+		Body:       pipeReader,
+		GetBody: func() (io.ReadCloser, error) {
+			// GetBody is called by the http client on request retries.
+			// We need to return a reader that will read from the buffer
+			// and then the pipe reader.
+			buf := writeBuffer.getBytes()
+			if buf == nil {
+				// Buffer is gone, so we can't return a reader.
+				return nil, fmt.Errorf("GetBody called after request was sent")
+			}
+			bufferReader := bytes.NewReader(buf)
+			return io.NopCloser(
+				io.MultiReader(bufferReader, pipeReader),
```

Comment:
Multireader will pull writes back into the buffer if needed for the next `GetBody`.